grafana/grafana
# loki: backend-queries: set interval-value in dataframe

With Loki, when running metric queries through the backend, we need to set in the value `interval` in the dataframe-field-config. this allows the grafana visualizations (timeseries etc.) to know where to "stop connecting points" and leave gaps.

(in loki-frontend-mode this happens here: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/result_transformer.ts#L499)

the change is minimal:
1. we extract the two fields into separate variables because we need to set an attribute on the time-field.
2. we set the interval-config attribute

about the diff's dataframe-snapshot-changes:
- because we changed something in the dataframes that we output, the dataframe-test-snapshots have changed
- in a snapshot, there is first a human-readable "summary" of the dataframe, and below the whole dataframe's bytes are shown, base64 encoded
- unfortunately, dataframe-field-configs are not included in the human-readable "summary" (we'll try to change this in https://github.com/grafana/grafana-plugin-sdk-go/pull/274), so the change we did here is not visible in the human-readable "summary" part. the base64 bytes have changed.
- there is no way to verify that the snapshots are correct, but i can offer two arguments for that:
    1. the only code-change is the new config-value, so that should be what caused the bytes in the snapshot to change
    2. the human-readable "summary" in the snapshots did not change (they are hidden by default on github), so those things (number of fields, content of fields) are unchanged in the dataframes


how to test:
- you need a Loki datasource
- you need to produce a graph that has a "gap" in the middle. one way is:
    - push some logs to Loki
    - wait 45minutes
    - push some more logs to Loki
    - now run a query like `count_over_time({level="info"}[10s])`, and observe that there is a gap between the two "mountains"
        - verify that the query is going through the backend-mode by observing the AJAX query in the browser-developer-tools
          - it should go to `/api/ds/query`, not to `/api/datasources/proxy/...`
          - if this is not happening, make sure that:
            - you have enabled the feature-flag `lokiBackendMode`
            - you are only running a single metric query on the `explore` page
        - on the main-branch the two mountains would be connected
